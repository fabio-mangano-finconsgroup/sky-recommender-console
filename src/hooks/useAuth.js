import React from 'react';
import { authContext } from '../auth/AuthProvider';

export default function useAuth() {
  return React.useContext(authContext);
}
