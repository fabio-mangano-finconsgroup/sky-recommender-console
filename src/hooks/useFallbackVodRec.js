import useSWR from 'swr';
import { getFallbackVodRec } from '../providers/vod-rec-provider/VodRecProvider';

export default function useFallbackVodRec() {
  return useSWR('http://localhost:3001/fallback-vod-recommendation', (url) =>
    getFallbackVodRec(url),
  );
}
