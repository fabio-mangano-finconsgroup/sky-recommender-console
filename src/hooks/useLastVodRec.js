import useSWR from 'swr';
import { getLastVodRec } from '../providers/vod-rec-provider/VodRecProvider';

export default function useLastVodRec(recVodId) {
  return useSWR(
    recVodId ? null : `http://localhost:3001/last-vod-recommendation`,
    (url) => getLastVodRec(url),
    {
      shouldRetryOnError: false,
      revalidateIfStale: false,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    },
  );
}
