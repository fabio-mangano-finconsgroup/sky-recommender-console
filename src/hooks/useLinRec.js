import useSWR from 'swr';
import { getLinRec } from '../providers/lin-rec-provider/LinRecProvider';

export default function useLinRec(recLinId) {
  return useSWR(recLinId ? `lin-vod/${recLinId}` : null, () =>
    getLinRec(recLinId),
  );
}
