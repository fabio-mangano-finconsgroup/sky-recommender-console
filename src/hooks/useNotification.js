import React from 'react';
import { AlertContext } from '../notifications/AlertProvider';

export default function useNotification() {
  const {
    actions: { addAlert, removeAlert },
    state,
  } = React.useContext(AlertContext);

  return { state, addAlert, removeAlert };
}
