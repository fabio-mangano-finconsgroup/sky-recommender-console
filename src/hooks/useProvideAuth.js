import React from 'react';
import { signIn, signOut } from '../providers/auth/AuthProvider';

export function useProvideAuth() {
  const [isAuth, setIsAuth] = React.useState(false);

  const signin = async () => {
    await signIn();
    setIsAuth(true);
  };

  const signout = async () => {
    await signOut();
    setIsAuth(false);
  };

  return {
    isAuth,
    signin,
    signout,
  };
}
