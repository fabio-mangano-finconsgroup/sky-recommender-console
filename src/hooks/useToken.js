import { useEffect, useState } from 'react';

export default function useToken() {
  const [token, setToken] = useState();

  useEffect(() => {
    const timeoutConnection = setInterval(() => {
      setToken(undefined);
    }, 7200000);
    return () => clearInterval(timeoutConnection);
  }, [setToken]);

  const saveToken = (userToken) => {
    setToken(userToken);
  };

  const removeToken = () => {
    setToken(undefined);
  };
  return {
    saveToken: saveToken,
    removeToken: removeToken,
    token,
  };
}
