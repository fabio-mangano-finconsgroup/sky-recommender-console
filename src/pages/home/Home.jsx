import React from 'react';
import AppBar from '@mui/material/AppBar';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import Container from '@mui/material/Container';
import Schedule from '../../components/schedule/Schedule';
import Fallback from '../../components/fallback/Fallback';
import Marginer from '../../components/marginer/Marginer';
import Box from '@mui/material/Box';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import { TabIcon, Warning } from './Home.styled';
import useFallbackVodRec from '../../hooks/useFallbackVodRec';
import { isExpired } from '../../utils/date';

const Home = () => {
  const [selectedTab, setSelectedTab] = React.useState(0);
  const [alertFallback, setAlertFallback] = React.useState(0);
  const { data: fallbackVodRec, error: fallbackVodRecError } =
    useFallbackVodRec();

  const handleChange = (event, newValue) => {
    setSelectedTab(newValue);
  };

  const handleAlertFallback = (recFallback) => {
    //todo quando ci sar√† il BE da rivedere
    let isExpiredFlag = 0;
    let rec = Array.isArray(recFallback)
      ? recFallback[0].recommendation
      : recFallback.recommendation;
    for (var i = 0; i < rec.length; i++) {
      if (isExpired(rec[i].endProgram)) {
        isExpiredFlag = 1;
      }
    }
    setAlertFallback(isExpiredFlag);
  };

  React.useEffect(() => {
    if (fallbackVodRec) {
      handleAlertFallback(fallbackVodRec);
    }
  }, [fallbackVodRec]);

  return (
    <>
      <AppBar data-test="app-bar" position="static">
        <Container maxWidth="xl">
          <Toolbar disableGutters>
            <Typography
              variant="h6"
              noWrap
              component="div"
              sx={{ mr: 2, display: { xs: 'none', md: 'flex' } }}
            >
              SKY Recommender Console
            </Typography>
          </Toolbar>
        </Container>
      </AppBar>
      <Box sx={{ m: 2 }} />
      <Container maxWidth="xl">
        <Marginer direction="horizontal" margin={10} />
        <Tabs data-test="nav-tabs" value={selectedTab} onChange={handleChange}>
          <Tab label="Schedule" />
          {alertFallback === 0 && (
            <Tab data-test="fallback-nav-tab" label="Fallback" />
          )}
          {alertFallback === 1 && (
            <TabIcon
              data-test="fallback-nav-tab"
              icon={<Warning />}
              iconPosition="end"
              label="Fallback"
            />
          )}
        </Tabs>

        {selectedTab === 0 && <Schedule />}
        {selectedTab === 1 && (
          <Fallback
            fallbackVodRec={fallbackVodRec}
            fallbackVodRecError={fallbackVodRecError}
            handleAlertFallback={handleAlertFallback}
          />
        )}
      </Container>
    </>
  );
};
export default Home;
