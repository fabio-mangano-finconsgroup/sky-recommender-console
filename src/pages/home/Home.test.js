import { render, waitFor, screen, fireEvent } from '@testing-library/react';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import * as useFallbackVodRec from '../../hooks/useFallbackVodRec';
import Home from './Home';

let mockedAddAlert = jest.fn();
jest.mock('../../hooks/useNotification', () => () => ({
  addAlert: mockedAddAlert,
}));

// Mock implementation of a sub component of the component being tested
jest.mock('../../components/schedule/Schedule', () => () => <div></div>);
jest.mock('../../components/fallback/Fallback', () => () => <div></div>);

describe('Home', () => {
  const removeToken = jest.fn();
  const props = {
    removeToken,
  };
  beforeAll(() => {
    jest.restoreAllMocks();
  });

  it('should render', async () => {
    const mockedUseFallbackVodRec = jest.fn(() => {
      return {
        data: { items: [{ id: '', type: 'FALLBACK', recommendation: [] }] },
        error: undefined,
      };
    });
    jest
      .spyOn(useFallbackVodRec, 'default')
      .mockImplementation(mockedUseFallbackVodRec);

    render(
      <LocalizationProvider dateAdapter={AdapterDateFns}>
        <Home {...props} />
      </LocalizationProvider>,
    );

    expect(document.body.childNodes[0].children).toMatchSnapshot();
  });

  it('should disconnect user if click on logout Icon', async () => {
    const mockedUseFallbackVodRec = jest.fn(() => {
      return {
        data: { items: [{ id: '', type: 'FALLBACK', recommendation: [] }] },
        error: undefined,
      };
    });
    jest
      .spyOn(useFallbackVodRec, 'default')
      .mockImplementation(mockedUseFallbackVodRec);

    render(
      <LocalizationProvider dateAdapter={AdapterDateFns}>
        <Home {...props} />
      </LocalizationProvider>,
    );

    const logoutButton = screen.getByTestId('LogoutIcon');
    await waitFor(() => fireEvent.click(logoutButton));

    expect(props.removeToken).toHaveBeenCalledTimes(1);
  });

  describe('Alert fallback icon', () => {
    it('should appear an warning icon if vod has warning message if rec object', async () => {
      const mockedUseFallbackVodRec = jest.fn(() => {
        return {
          data: {
            items: [
              {
                id: '',
                type: 'FALLBACK',
                recommendation: [{ warningMessage: 'out of date' }],
              },
            ],
          },

          error: undefined,
        };
      });
      jest
        .spyOn(useFallbackVodRec, 'default')
        .mockImplementation(mockedUseFallbackVodRec);

      render(
        <LocalizationProvider dateAdapter={AdapterDateFns}>
          <Home {...props} />
        </LocalizationProvider>,
      );

      await waitFor(() => {
        const warningIcon = screen.queryByTestId('warning-fallback');
        expect(warningIcon).toBeInTheDocument();
      });
    });

    it('should not appear an warning icon if vod has not warning message if rec object', async () => {
      const mockedUseFallbackVodRec = jest.fn(() => {
        return {
          data: {
            items: [
              {
                id: '',
                type: 'FALLBACK',
                recommendation: [{ warningMessage: '' }],
              },
            ],
          },

          error: undefined,
        };
      });
      jest
        .spyOn(useFallbackVodRec, 'default')
        .mockImplementation(mockedUseFallbackVodRec);

      render(
        <LocalizationProvider dateAdapter={AdapterDateFns}>
          <Home {...props} />
        </LocalizationProvider>,
      );

      await waitFor(() => {
        const warningIcon = screen.queryByTestId('warning-fallback');
        expect(warningIcon).not.toBeInTheDocument();
      });
    });
  });
});
