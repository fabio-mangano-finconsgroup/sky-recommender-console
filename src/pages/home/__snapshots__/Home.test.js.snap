// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render 1`] = `
HTMLCollection [
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    data-test="app-bar"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular css-1cndkqf-MuiToolbar-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
        >
          SKY Recommender Console
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-199uyvf-MuiSvgIcon-root"
            data-testid="LogoutIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>,
  <div
    class="MuiBox-root css-1c5ij41"
  />,
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
  >
    <span
      class="css-yviemf"
      direction="horizontal"
      margin="10"
    />
    <div
      class="MuiTabs-root css-1ujnqem-MuiTabs-root"
      data-test="nav-tabs"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Schedule
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            data-test="fallback-nav-tab"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Fallback
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div />
  </div>,
]
`;
